extensions/amp-story-shopping/0.1/amp-story-shopping-tag.js

```diff
@@ -1,30 +1,54 @@
-import {Layout, applyFillContent} from '#core/dom/layout';
+import {Layout} from '#core/dom/layout';
 
-const TAG = 'amp-story-shopping-tag';
+import {Services} from '#service';
+
+import {devAssert} from '#utils/log';
+
+import {StateProperty} from '../../amp-story/1.0/amp-story-store-service';
 
 export class AmpStoryShoppingTag extends AMP.BaseElement {
   /** @param {!AmpElement} element */
   constructor(element) {
     super(element);
 
-    /** @private {string} */
-    this.myText_ = TAG;
-
-    /** @private {?Element} */
-    this.container_ = null;
+    /** @private @const {?./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = null;
   }
 
   /** @override */
   buildCallback() {
-    super.buildCallback();
-    this.container_ = this.element.ownerDocument.createElement('div');
-    this.container_.textContent = this.myText_;
-    this.element.appendChild(this.container_);
-    applyFillContent(this.container_, /* replacedContent */ true);
+    return Services.storyStoreServiceForOrNull(this.win).then(
+      (storeService) => {
+        devAssert(storeService, 'Could not retrieve AmpStoryStoreService');
+        this.storeService_ = storeService;
+      }
+    );
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.storeService_.subscribe(
+      StateProperty.SHOPPING_DATA,
+      (shoppingState) => this.updateShoppingTag_(shoppingState),
+      true /** callToInitialize */
+    );
   }
 
   /** @override */
   isLayoutSupported(layout) {
     return layout === Layout.CONTAINER;
   }
+
+  /**
+   * @param {!Object} shoppingState
+   * @private
+   */
+  updateShoppingTag_(shoppingState) {
+    const tagData = shoppingState[this.element.getAttribute('data-tag-id')];
+    if (tagData) {
+      this.measureElement(() => {
```

Comment:
This should be updated to `mutateElement()` since `textContent` is being mutated. (The previous `measureElement()` suggestion was accidental)

Reply:
that makes more sense. Thanks Corey!